src/NHibernate/Linq/Visitors/SelectClauseNominator.cs

```diff
@@ -109,6 +112,29 @@ public override Expression VisitExpression(Expression expression)
 			return expression;
 		}
 
+		private bool DbCanCast(System.Type type)
```

Comment:
I think that probably needs to be decided by Driver/Dialect, as, for example, MySQL support only few types in cast function. 

https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Dialect/MySQLDialect.cs#L317
http://dev.mysql.com/doc/refman/5.7/en/cast-functions.html